src/environmentd/src/deployment/preflight.rs

```diff
@@ -132,27 +131,35 @@ pub async fn preflight_0dt(
         mut openable_adapter_storage,
         catalog_metrics,
     }: PreflightInput,
-) -> Result<(Box<dyn OpenableDurableCatalogState>, bool), anyhow::Error> {
+) -> Result<
+    (
+        Box<dyn OpenableDurableCatalogState>,
+        bool,
+        Option<trigger::Trigger>,
+    ),
+    anyhow::Error,
+> {
     info!(%deploy_generation, "performing 0dt preflight checks");
 
     if !openable_adapter_storage.is_initialized().await? {
         info!("catalog not initialized; booting as leader with writes allowed");
         deployment_state.set_is_leader();
-        return Ok((openable_adapter_storage, false));
+        return Ok((openable_adapter_storage, false, None));
     }
 
     let catalog_generation = openable_adapter_storage.get_deployment_generation().await?;
     info!(%catalog_generation, "catalog initialized");
     if catalog_generation < deploy_generation {
         info!("this deployment is a new generation; booting in read only mode");
 
+        let (clusters_hydrated_trigger, clusters_hydrated_receiver) = trigger::channel();
+
         // Spawn a background task to handle promotion to leader.
         mz_ore::task::spawn(|| "preflight_0dt", async move {
-            // Simulate waiting for readiness to promote by waiting for a
-            // fixed 30s.
-            // TODO(benesch): actually wait for clusters to hydrate.
-            info!("waiting to be ready to promote");
-            time::sleep(Duration::from_secs(30)).await;
+            info!("waiting to be ready to promote, waiting for clusters to be hydrated");
```

Comment:
feel like you can go for the simple thing here:

```suggestion
            info!("waiting for clusters to be hydrated");
```